src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1397,83 @@ struct OptimizeInstructions
     return binary;
   }
 
+  Expression* optimizeBulkMemory(Expression* curr) {
+    if (auto* memCopy = curr->dynCast<MemoryCopy>()) {
+      FeatureSet features = getModule()->features;
+
+      auto destHasSideEffects =
+        EffectAnalyzer(getPassOptions(), features, memCopy->dest)
+          .hasSideEffects();
+
+      // memory.copy(x, x, sz)  ==>  nop
+      if (!destHasSideEffects &&
+          ExpressionAnalyzer::equal(memCopy->dest, memCopy->source)) {
+        return ExpressionManipulator::nop(curr);
+      }
+      // memory.copy(dst, src, C)  ==>  store(dst, load(src))
+      if (auto* csize = memCopy->size->dynCast<Const>()) {
+        auto bytes = csize->value.geti32();
+
+        switch (bytes) {
+          case 0: {
+            if (!destHasSideEffects &&
+                !EffectAnalyzer(getPassOptions(), features, memCopy->source)
+                   .hasSideEffects()) {
+              return ExpressionManipulator::nop(curr);
+            }
+            break;
+          }
+          case 1:
+          case 2:
+          case 4: {
+            Builder builder(*getModule());
+            return builder.makeStore(
+              bytes, // bytes
+              0,     // offset
+              bytes, // align
+              memCopy->dest,
+              builder.makeLoad(
+                bytes, false, 0, bytes, memCopy->source, Type::i32),
+              Type::i32);
+          }
+          case 8: {
+            Builder builder(*getModule());
+            return builder.makeStore(
+              bytes, // bytes
+              0,     // offset
+              bytes, // align
+              memCopy->dest,
+              builder.makeLoad(
+                bytes, false, 0, bytes, memCopy->source, Type::i64),
+              Type::i64);
+          }
+          case 16: {
+            if (getPassOptions().shrinkLevel == 0) {
+              // this increase add extra 2 bytes so apply it only for
+              // minimal shrink level
```

Comment:
Don't all of these transformations increase code size?

Reply:
No. I specially checked this.

Reply:
But a single `memory.copy` instruction is 4 bytes (prefix + opcode + 2 memory indices) and loads and stores are three bytes each (opcode + alignment + offset). So if I'm counting correctly, these optimizations normally increase code size by 2 bytes and the SIMD one increases code size by 4 bytes due to the two extra prefix bytes.

Reply:
```wat
(module
  (memory $0 0)
  (func (export "a") (param i32 i32)
    (memory.copy
      (local.get 0)
      (local.get 1)
      (i32.const 4)
    )
  )
)
```
will take `64` bytes
```wat
(module
  (memory $0 0)
  (func (export "a") (param i32 i32)
    (i32.store align=1 
      (local.get 0) 
      (i32.load align=1 (local.get 1))
    )
  )
)
```
will take also the same `64` bytes
I tested via [online wat2wasm](https://webassembly.github.io/wabt/demo/wat2wasm/)

Reply:
> single memory.copy instruction is 4 bytes (prefix + opcode + 2 memory indices)

Also one instruction for byte size

Reply:
Ah, right. I forgot to account for the size parameter that gets optimized out. Thanks for the reminder!